src/test/cpp/src/test_genomicsdb_api.cc

```diff
@@ -0,0 +1,239 @@
+}
+
+TEST_CASE("api query_variants direct", "[query_variants]") {
+  GenomicsDB* gdb = new GenomicsDB(workspace, callset_mapping, vid_mapping, reference_genome, {"DP"}, 40);
+  check_query_variants_results(gdb, array, gdb->query_variants(array, {{0,1000000000}}, {{0,3}}));
+  check_query_variants_results(gdb, array, gdb->query_variants(array, {{0,1000000000}}));
+  check_query_variants_results(gdb, array, gdb->query_variants(array));
+  delete gdb;
+}
+
+TEST_CASE("api query_variants with json", "[query_variants") {
+  GenomicsDB* gdb = new GenomicsDB(query_json, loader_json);
+  check_query_variants_results(gdb, array, gdb->query_variants());
+  delete gdb;
+}
+
+
+class NullVariantCallProcessor : public GenomicsDBVariantCallProcessor {
+ public:
+  NullVariantCallProcessor() {
+  };
+
+  void process(interval_t interval) {
+  };
+
+  void process(uint32_t row,
+               genomic_interval_t genomic_interval,
+               std::vector<genomic_field_t> genomic_fields){
+  };
+};
+
+class OneQueryIntervalProcessor : public GenomicsDBVariantCallProcessor {
+ public:
+  OneQueryIntervalProcessor() {};
+
+  ~OneQueryIntervalProcessor() {
+    CHECK(m_num_query_intervals == 1);
+    CHECK(m_processed_rows == 5);
+  }
+
+  void process(interval_t interval) {
+    m_num_query_intervals++;
+    CHECK(interval.first == 0);
+    CHECK(interval.second == 1000000000);
+  };
+
+  void process(uint32_t row,
+               genomic_interval_t genomic_interval,
+               std::vector<genomic_field_t> genomic_fields){
+    m_processed_rows++;
+    CHECK(row <= 2);
+    if (row == 2) {
+      CHECK(genomic_interval.contig_name == "1");
+      CHECK(genomic_interval.interval.first == 17385);
+      CHECK(genomic_interval.interval.second == 17385);
+      CHECK(genomic_fields.size() == 3);
+      CHECK(genomic_fields[0].first == "REF");
+      CHECK(genomic_fields[0].second == "\"G\"");
+      CHECK(genomic_fields[1].first == "ALT");
+      CHECK(genomic_fields[1].second == "[ \"A\", \"<NON_REF>\" ]");
+      CHECK(genomic_fields[2].first == "DP");
+      CHECK(genomic_fields[2].second == "76");
+    }
+  };
+
+  int m_num_query_intervals = 0;
+  int m_processed_rows = 0;
+};
+
+class TwoQueryIntervalsProcessor : public GenomicsDBVariantCallProcessor {
+ public:
+  TwoQueryIntervalsProcessor() {};
+
+  ~TwoQueryIntervalsProcessor() {
+    CHECK(m_num_query_intervals == 2);
+    CHECK(m_processed_rows == 5);
+  }
+
+  void process(interval_t interval) {
+    m_num_query_intervals++;
+    if (m_num_query_intervals == 1) {
+      CHECK(interval.first == 0);
+      CHECK(interval.second == 17000);
+    } else {
+      CHECK(interval.first == 17000);
+      CHECK(interval.second == 18000);
+    }
+  };
+
+  void process(uint32_t row,
+               genomic_interval_t genomic_interval,
+               std::vector<genomic_field_t> genomic_fields){
+    m_processed_rows++;
+  };
+
+  int m_num_query_intervals = 0;
+  int m_processed_rows = 0;
+};
+
```

Comment:
I'm sure there's an endless list of things we could try to test here...but maybe also add a test case that exercises the rank part of this to make sure that if someone does try to throw more threads at this, we know it'll work.

Reply:
Yes, I will add a test with rank.